### PR TITLE
Inline svg broken

### PR DESCRIPTION
**Pull request by Vaelatern**
It's broken.

I need to bubble up the error.

IDK how.

#8 

Closures keep surrounding me and preventing errors from bubbling up.

### CODE REVIEW

**Comment by danieleades** [comment]
Not to worry, we'll go on this little journey together.

First of all, I'm probably not looking to remove the current implementation, unless the inline SVGs can meet all the same use cases. For example the current implementation allows the document to be built as a PDF. If the SVGs are fully-inlined, is that still supported? If everything worked the same, then I'd definitely reconsider that position. So the solution is to create a new module. You could probably start by copying and pasting, but we'll want to find some ways to reduce the duplication before this is merged.

You'll notice I've got some cheeky unwraps in the original. That's definitely not best practice for error handling, and something I was planning to improve. That said, if the filesystem isn't read/writeable then the user has bigger problems than this humble little plugin. It's generally not too hard to add error handling after that fact, so don't be afraid to unwrap a few errors while you're prototyping.

I'll give this a proper review when I'm on a laptop, but it looks a decent start

**Comment by Vaelatern** [comment]
"First of all, I'm probably not looking to remove the current implementation, unless the inline SVGs can meet all the same use cases. For example the current implementation allows the document to be built as a PDF. If the SVGs are fully-inlined, is that still supported?"

D2 should be compiled to png if it's going to be used non-interactively like this, as of version 0.1.4

As for removal, I'm trying to make it work at all. Then the plumbing can be installed to make it two compilers in one.

File: src/lib.rs [hunk 1]
```diff
@@ -50,25 +50,23 @@ impl Preprocessor for D2 {
                 cmark(events, &mut buf).unwrap();
                 chapter.content = buf;
             }
-        });
-
-        Ok(book)
+        })
```
Comment (danieleades) [inline src/lib.rs:53]: right now, your `Preprocessor::run` method isn't returning anything, but it needs to return a `Result<Book, Error>`.

try adding an `Ok(book)` statement at the end (which is itself shorthand for `Result::<Book, Error>::Ok(book)`)

File: src/lib.rs [hunk 2]
```diff
@@ -50,25 +50,23 @@ impl Preprocessor for D2 {
                 cmark(events, &mut buf).unwrap();
```
Comment (danieleades) [inline src/lib.rs:50]: the 'process_events' method is return a Result<impl Iterator, Error>, but the `cmark` method is expecting an `impl Iterator` (not wrapped in Result). You need to get the value from inside the `Result` by either handling or unwrapping the potential error value.
You can do this with the 'try' operator ('?') if the error type can be converted to the error type of the 'run' method. If not, you'll have to handle it some other way. I won't tell anyone if you want to just explicitly 'unwrap' it for now while you get the rest working and then come back and revisit it later

File: src/lib.rs [hunk 3]
```diff
@@ -50,25 +50,23 @@ impl Preprocessor for D2 {
...
+) -> Result<impl Iterator<Item = Event<'a>> + 'a, Error> {
     let mut in_block = false;
     // if Windows crlf line endings are used, a code block will consist
     // of many different Text blocks, thus we need to buffer them in here
     // see https://github.com/raphlinus/pulldown-cmark/issues/507
     let mut diagram = String::new();
     let mut diagram_index = 0;
 
-    events.flat_map(move |event| {
+    Ok(events.try_for_each(move |event| {
```
Comment (danieleades) [inline src/lib.rs:69]: try_for_each isn't what you're looking for here. `try_for_each` will run the closure on each element in the iterator, and then return nothing. What you want is to do is process each element in the iterator, and then pass it along. You'll want to use the 'map' family of iterator adaptors.

I've used `flat_map` in the original implementation. It takes an iterator, runs each element through a closure that returns an iterator, and then 'flattens' the inner iterators by chaining them all together

Have a read through the iterator docs - https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html

File: src/lib.rs (hunk 2 above)
Comment (Vaelatern) [reply src/lib.rs:50]: Ah so I can't use ? here because it's in a closure that returns `()`

File: src/lib.rs (hunk 1 above)
Comment (Vaelatern) [reply src/lib.rs:53]: Cool, done that

File: src/lib.rs (hunk 3 above)
Comment (Vaelatern) [reply src/lib.rs:69]: I reverted to flat_map because none of the other iterators listed seemed to allow me to return a Result

**Comment by Vaelatern** [comment]
```
warning: unused imports: `LinkType`, `Tag`
 --> src/backend.rs:9:37
  |
9 | use pulldown_cmark::{CowStr, Event, LinkType, Tag};
  |                                     ^^^^^^^^  ^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error[E0277]: the `?` operator can only be used in a closure that returns `Result` or `Option` (or another type that implements `FromResidual`)
  --> src/lib.rs:94:62
   |
68 |       events.flat_map(move |event| {
   |  _____________________-
69 | |         match (&event, in_block) {
70 | |             // check if we are entering a d2 codeblock
71 | |             (
...  |
94 | |                 vec![backend.render(render_context, &diagram)?]
   | |                                                              ^ cannot use the `?` operator in a closure that returns `Vec<Event<'_>>`
...  |
98 | |         }
99 | |     })
   | |_____- this function should return `Result` or `Option` to accept `?`
   |
   = help: the trait `FromResidual<Result<Infallible, mdbook::errors::Error>>` is not implemented for `Vec<Event<'_>>`

error[E0308]: mismatched types
  --> src/lib.rs:68:5
   |
60 |   ) -> Result<impl Iterator<Item = Event<'a>> + 'a, Error> {
   |        --------------------------------------------------- expected `Result<_, mdbook::errors::Error>` because of return type
...
68 | /     events.flat_map(move |event| {
69 | |         match (&event, in_block) {
70 | |             // check if we are entering a d2 codeblock
71 | |             (
...  |
98 | |         }
99 | |     })
   | |______^ expected enum `Result`, found struct `FlatMap`
   |
   = note: expected enum `Result<_, mdbook::errors::Error>`
            found struct `FlatMap<impl Iterator<Item = Event<'a>> + 'a, Vec<Event<'_>>, [closure@src/lib.rs:68:21: 68:33]>`
help: try wrapping the expression in `Ok`
   |
68 ~     Ok(events.flat_map(move |event| {
69 |         match (&event, in_block) {
 ...
98 |         }
99 ~     }))
   |

error[E0277]: `Result<impl Iterator<Item = Event<'_>>, mdbook::errors::Error>` is not an iterator
   --> src/lib.rs:48:23
    |
48  |                 cmark(events, &mut buf).unwrap();
    |                 ----- ^^^^^^ `Result<impl Iterator<Item = Event<'_>>, mdbook::errors::Error>` is not an iterator
    |                 |
    |                 required by a bound introduced by this call
    |
    = help: the trait `Iterator` is not implemented for `Result<impl Iterator<Item = Event<'_>>, mdbook::errors::Error>`
note: required by a bound in `cmark`
   --> /home/toyam/.cargo/registry/src/github.com-1ecc6299db9ec823/pulldown-cmark-to-cmark-10.0.4/src/lib.rs:627:8
    |
627 |     I: Iterator<Item = E>,
    |        ^^^^^^^^^^^^^^^^^^ required by this bound in `cmark`

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
```

**Comment by Vaelatern** [comment]
Sadly I can't just unwrap it... for some reason the d2 reader is not giving me anything, so there is a problem with my reading the file, somehow, probably due to where it is on the filesystem.

**Comment by danieleades** [comment]
@Vaelatern i've done a quick job of implementing this in #17

**Comment by danieleades** [comment]
i'm going to close this since it's inactive, and was implemented in #17

**Comment by Vaelatern** [comment]
Thanks! I went inactive on it because I wrote github.com/Vaelatern/mdbook-d2-go to fit the use case

**Pull request closed without merging**